crates/polars-io/src/file_cache/entry.rs

```diff
@@ -0,0 +1,351 @@
+                "[file_cache::entry] try_open_assume_latest: did not find cached file for uri = {}",
+                self.uri.clone()
+            );
+        }
+
+        self.try_open_check_latest()
+    }
+
+    fn try_open_check_latest(&mut self) -> PolarsResult<std::fs::File> {
+        let verbose = config::verbose();
+        let remote_metadata = self.file_fetcher.fetch_metadata()?;
+        let cache_guard = GLOBAL_FILE_CACHE_LOCK.lock_any();
+
+        {
+            let metadata_file = &mut self.metadata.acquire_shared().unwrap();
+            update_last_accessed(metadata_file);
+
+            if let Ok(metadata) = self.try_get_metadata(metadata_file, &cache_guard) {
+                if metadata.remote_last_modified == remote_metadata.last_modified
+                    && metadata.local_size == remote_metadata.size
+                {
+                    let data_file_path = self.get_cached_data_file_path();
+                    let cmp_result = metadata.compare_local_state(data_file_path);
+
+                    if cmp_result.is_ok() {
+                        if verbose {
+                            eprintln!("[file_cache::entry] try_open_check_latest: opening already fetched file for uri = {}", self.uri.clone());
+                        }
+                        return Ok(finish_open(data_file_path, metadata_file));
+                    }
+                }
+            }
+        }
+
+        let metadata_file = &mut self.metadata.acquire_exclusive().unwrap();
+        let metadata = self
+            .try_get_metadata(metadata_file, &cache_guard)
+            .unwrap_or_else(|_| Arc::new(EntryMetadata::new_with_uri(self.uri.clone())));
+
+        if metadata.remote_last_modified == remote_metadata.last_modified
+            && metadata.local_size == remote_metadata.size
+        {
+            let data_file_path = self.get_cached_data_file_path();
+            let cmp_result = metadata.compare_local_state(data_file_path);
+
+            if cmp_result.is_ok() {
+                if verbose {
+                    eprintln!(
+                        "[file_cache::entry] try_open_check_latest: opening already fetched file (lost race) for uri = {}",
+                        self.uri.clone()
+                    );
+                }
+                return Ok(finish_open(data_file_path, metadata_file));
+            }
+        }
+
+        if verbose {
+            eprintln!(
+                "[file_cache::entry] try_open_check_latest: fetching new data file for uri = {}, remote_last_modified = {}, remote_size = {}",
+                self.uri.clone(),
+                remote_metadata.last_modified,
+                remote_metadata.size
+            );
+        }
+
+        let data_file_path = &get_data_file_path(
+            self.path_prefix.to_str().unwrap().as_bytes(),
+            self.uri_hash.as_bytes(),
+            remote_metadata.last_modified,
+        );
+        let _ = std::fs::remove_file(data_file_path);
```

Comment:
This will block until a new file is downloaded right?

Reply:
hmm, not exactly..- this was for in case there was a partial download from an aborted process. I've also left a comment above it.